frontend/javascript/components/postupload/post-upload.vue

```diff
@@ -0,0 +1,1494 @@
+    gotoStep()
+    return
+  }
+  documentsImagesConverting.value = true
+  // XXX wild mix of Vue2&3 refs, not the cleanest, but a bus was basically impossible
+  // we are calling a grand-child component's method
+  documentUploader.value.$refs.imageDocument?.[0]?.convertImages?.()
+  // ^child              ^vue2 ^grandchild...REN!  ^method
+  // alternative: pass a prop, watch it, react?
+}
+const documentsImagesDocumentFilenameDefault = 'brief.pdf'
+const documentsImagesDocumentFilename = ref(
+  documentsImagesDocumentFilenameDefault
+)
+const documentsImagesDocumentFilenameNormalized = computed(() =>
+  documentsImagesDocumentFilename.value.replace(/\.pdf$/, '')
+)
+
+const documentsBasicOperations = ref(false)
+
+const pdfRedaction = ref()
+const pdfRedactionCurrentHasRedactions = ref(false)
+const pdfRedactionProcessing = ref(false)
+const pdfRedactionRedact = () => {
+  pdfRedactionProcessing.value = true
+  // XXX calling child's method
+  // alternatively, could listen to an event
+  pdfRedaction.value
+    .redactOrApprove()
+    // .then(delay(3000))
+    .then(() => {
+      pdfRedactionProcessing.value = false
+      pdfRedactionCurrentHasRedactions.value = false
+      gotoStep()
+    })
+}
+const pdfRedactionUploaded = () => {
+  // TODO handle errors?
+  // XXX again, we're calling child's method, could handle attachments here, store-like,
+  //   and pass as a "override prop" to the components
+  documentUploader.value.refreshAttachments()
+}
+
+const documentUploaderSelectAll = (val) => {
+  documentUploader.value.setAllSelect(val)
+}
+
+const scrollHeaderIntoViewIfNecessary = () => {
+  // we could bail early if !isDesktop.value,
+  // but this check fill automagically still work as the header is always in viewport on mobile
+  // alt: check appshell-header-container--desktop, if hidden all rect attributes == 0
+  const header = document.querySelector('.appshell-header')
+  const rect = header.getBoundingClientRect()
+  if (rect.bottom < 0) {
+    // timeout to make it feel less jarring
+    setTimeout(() => header.scrollIntoView({ behavior: 'smooth' }), 500)
+  }
+}
+
+const mobileHeaderTitle2 = computed(() => {
+  switch (step.value) {
+    case 1100:
+      return i18n.value.addLetter
+    case 4570:
+      return 'Fertig'
+  }
+  if (step.value < 2000) {
+    return 'Brief hochladen oder scannen'
+  }
+  if (2000 <= step.value && step.value < 3000) {
+    return 'Infos eingeben'
+  }
+  if (3000 <= step.value && step.value < 4000) {
+    return 'Schwärzen'
+  }
+  if (step.value >= 4000) {
+    return 'Vorschau'
+  }
+  return '(Untertitel)'
+})
+
+const progress = computed(() => {
+  // roughly map to {0, ½, 1}
+  return Math.min(Math.floor((step.value - 1000) / 1000), 2) / 2
+})
+
+const progressStep = computed(() =>
+  // map to {0, 1, 2}
+  Math.min(Math.floor((step.value - 1000) / 1000), 2)
+)
+
+const gotoStep = (nextStep) => {
+  stepHistory.push(nextStep || getNextStep())
+}
+
+const gotoValid = computed(() => {
```

Comment:
maybe we should use a proper state machine for this, like xstate? this is very confusing to me

Reply:
yes! having a visual representation will be very beneficial

Reply:
plan of attack:
1. attach the current implementation as-is to history api (like ` .../#1100'`)
2. see if xstate can a. handle our state machine part and b. work with hash-history
3. if not, consider a lightweight, handrolled state machine implementation

Reply:
maybe vue-router could also work for this

Reply:
but doesn't vue-router also always expose the state to the back-button, leaving states reachable which shouldn't be? resp., necessitate complex navigation guards, which are state-machine-like...